### PR TITLE
fv3fit.emulation: add TensorFactory abstraction

### PR DESCRIPTION
**Pull request by nbren12**
This refactor prepares for temperature scaled transformations. Since these tranforms will depend on data, I need to separate the config layer from the actual transform object. The config objects satisfy the `TransformFactory` interface and the transforms implement `TensorTranform`.

The diff is somewhat large because the `transform` object was coupled to the models, losses, and top-level config object...I tried to remove this coupling where it was easy.

### CODE REVIEW

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 1]
```diff
@@ -255,16 +263,20 @@ def main(config: TrainConfig, seed: int = 0):
...
+        config.train_url, config.nfiles, config.model_variables
     )
-    test_ds = nc_dir_to_tf_dataset(
-        config.test_url, config.get_dataset_convertor(), nfiles=config.nfiles_valid
+    test_ds = config.open_dataset(
+        config.test_url, config.nfiles_valid, config.model_variables
     )
 
     train_set = next(iter(train_ds.shuffle(100_000).batch(50_000)))
+    transform = config.build_transform(train_set)
```
Comment (nbren12) [inline external/fv3fit/fv3fit/train_microphysics.py:274]: need to only build the transform once.

File: external/fv3fit/fv3fit/emulation/models/transformed_model.py [hunk 2]
```diff
@@ -74,12 +72,18 @@ def build(
         # contains both inputs and outputs the serialized model will think its
         # outputs are also inputs and never be able to evaluate...even though
         # calling `model(data)` works just fine.
+        input_names = set(
```
Comment (nbren12) [inline external/fv3fit/fv3fit/emulation/models/transformed_model.py:75]: can't use the `backwards_names` method any more since that was moved to `TransformFactory`, but it is still possible to infer the input names using this logic.

File: external/fv3fit/fv3fit/emulation/models/transformed_model.py [hunk 3]
```diff
@@ -63,9 +63,7 @@ def build(
         factory = FieldFactory(
             self.timestep_increment_sec, self.normalize_key, self.enforce_positive, data
         )
-        model = TransformedModel(
```
Comment (nbren12) [inline external/fv3fit/fv3fit/emulation/models/transformed_model.py:66]: Removed `transform` from this low level object since it shouldn't need to know this.

File: external/fv3fit/fv3fit/emulation/transforms/transforms.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    def forward(self, x: TensorDict) -> TensorDict:
+        for transform in self.transforms:
+            try:
+                x = transform.forward(x)
+            except KeyError:
+                pass
+        return x
+
+    def backward(self, y: TensorDict) -> TensorDict:
+        for transform in self.transforms:
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/emulation/transforms/transforms.py:69]: Can you add a test for the composed transform that uses a couple non associative operations?

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
The refactor makes sense and removing the coupling in the model and loss looks great. 

I was kind of confused by the PR description that said "I need to separate the config layer from the actual transform object.". Is the intent that the config is still needed at least to use `config.build_transform` at the start but from then on the config is no longer needed to do stuff using the transform?

Just a couple minor requests about the composed transform. LMK if I am mistaken about those.

File: external/fv3fit/fv3fit/emulation/transforms/transforms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/transforms/transforms.py:69]: I hadn't considered this. I had been assuming the transforms are independent, but as you mention this is unnecessary. I will add a test.

**Comment by nbren12** [comment]
> I was kind of confused by the PR description that said "I need to separate the config layer from the actual transform object.". Is the intent that the config is still needed at least to use config.build_transform at the start but from then on the config is no longer needed to do stuff using the transform?

The basic struggle is that the Transform can no longer be represented in yaml config since it will need to depends on data. The `TransformFactory` is the yaml-able object.

File: external/fv3fit/fv3fit/emulation/models/transformed_model.py [hunk 5]
```diff
@@ -74,12 +72,18 @@ def build(
         # contains both inputs and outputs the serialized model will think its
         # outputs are also inputs and never be able to evaluate...even though
         # calling `model(data)` works just fine.
+        input_names = set(
+            transform.backward({key: data[key] for key in self.input_variables})
+        )
```
[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/models/transformed_model.py:77]: > if you are training a model with a transformed input, does input_variables refer to the untransformed variable name?
`input_variables` refers to the name in transformed space. `input_variables` isn't a configuration option of this class btw. It is inferred by looking at the list of `fields`. Here is an example configuration: https://github.com/ai2cm/fv3net/blob/af35a4adfc18d0c0925baeae08c73545c180a0c1/projects/microphysics/train/log-cloud.yaml#L48


> and then when predicting stuff with the trained model( input_dict) -> output_dict are the dict keys still the untransformed names, and the backwards names are an internal implementation detail?

There are two layers of "models" here. The inner ML model (with all the trainable weights) is in the transformed space. The transform transforms the orginal inputs (cloud, temperature, etc) so they can be passed to the inner model, and then backwards transforms any outputs.  This is all wrapped in an "outer" keras model which is what is trained with the data and-or serialized for a prognostic run.

[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/models/transformed_model.py:79]: > the backwards names are an internal implementation detail?
Basically. It is generally needed to figure out what variables the outer model (which will be trained/serialized) requires in untransformed space. This is needed either to 1) generate an appropriate list of `tf.keras.Input` objects of 2) request the right variables from the dataset.

File: external/fv3fit/fv3fit/emulation/models/transformed_model.py [hunk 6]
```diff
@@ -149,7 +151,9 @@ def build_architecture(
         return config.build(output_features)
 
 
-class TransformedModel(tf.keras.layers.Layer):
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/emulation/models/transformed_model.py:160]: Thanks, this makes it clearer.

File: external/fv3fit/fv3fit/emulation/models/transformed_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/emulation/models/transformed_model.py:79]: Thanks, that makes sense now. The `InnerModel` naming makes it more clear.

**Review by AnnaKwa — APPROVED** [review]
Thanks! LGTM.
